libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/VoiceInstructionLoader.java

```diff
@@ -0,0 +1,198 @@
+package com.mapbox.services.android.navigation.v5.navigation;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.google.auto.value.AutoValue;
+import com.mapbox.api.directions.v5.models.LegStep;
+import com.mapbox.api.directions.v5.models.VoiceInstructions;
+import com.mapbox.api.speech.v1.MapboxSpeech;
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import okhttp3.Cache;
+import okhttp3.ResponseBody;
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+@AutoValue
+public abstract class VoiceInstructionLoader {
+  private static final int NUMBER_TO_CACHE = 3;
+  private static VoiceInstructionLoader instance = null;
+
+  /**
+   * Returns the singleton instance of VoiceInstructionLoader. It must first be initialized through
+   * a builder.
+   *
+   * @return current instance if it's initialized, or null
+   */
+  public static synchronized VoiceInstructionLoader getInstance() {
+    return instance;
+  }
+
+  /**
+   * Makes the call to MapboxSpeech to get the given string instruction as a sound file.
+   *
+   * @param instruction text to dictate
+   * @param textType "ssml" or "text"
+   * @param callback to relay retrofit status
+   */
+  public void getInstruction(String instruction, String textType, Callback<ResponseBody> callback) {
+    getMapboxBuilder()
+      .instruction(instruction)
+      .textType(textType)
+      .build()
+      .enqueueCall(callback);
+  }
+
+  /**
+   * Makes call to MapboxSpeech with empty callbacks. This is so that the result is cached in the
+   * cache specified in the builder.
+   *
+   * @param routeProgress to get instructions from
+   * @param isFirst whether this is the first call. This way, if we're caching three ahead, on the
+   *                first call the first two instructions will also be cached.
+   */
+  public void cacheInstructions(RouteProgress routeProgress, boolean isFirst) {
+    List<VoiceInstructions> voiceInstructions = getNextInstructions(routeProgress);
+
+    if (isFirst) {
+      for (int i = 0; i <= NUMBER_TO_CACHE - 1; i++) {
```

Comment:
Minor thing. Could we extract `1` magic number into a constant called for example `OFFSET`?